### ISSUE TITLE
Merging tuples in templates looses labeling

### ISSUE THREAD

**Issue by brett19** [microsoft/TypeScript#43020]
# Bug Report

### 🔎 Search Terms

- labeled tuple
- tuple merging

### 🕗 Version & Regression Information

This is a bug in all versions of Typescript I tested which support labeled tuples, variadic templates and non-last rest parameters in tuples (4.2.x)

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.2.0-beta#code/PTAEGUHsFsFNQBawDYAdYCcDOAoALgJ7qgAiGkqAYgJbZ4A8AKqLAB56wB2AJlqAIacCAbQC6APlABeUMzYcefYdU4AzTKAD6AGlAA6AyvUZZo0AH5ZoAFyhOsAG6Z8ReABl+WBnPZdeAoTFJGR8Ff2EDPSMNHVBokzczSzcbO0dnHBAIGHgsWDwAV1QXYkZYL0oCzgBjaVAACn5bLwwVAHNdACNbetZbTgLoTsxdAma8Vs42gEppSQdIam5ZqXnF7hL4Mq8ABX4Mfmg+GT2DuA5sJnK8SprxHEywAHVIDABrPgB3ajwESAK8AJQBxoKhkPwOHpQKAAJrUFDhABetgAohhyBhRuNJm1RJtZNdGK4Xu9jjhoREDMJkaA0RjRLpImQKDQ6PRTod8pgsPQPF4rrt9ocsOJRXjHqAAEIYWD8D6gT5ITigAp5bgKn4IIEgsEQ2CWWHw5DhOmvXQtdp4wilQmuRhgpgsXyKUCUvSCEQMhqRfZtLC2D0rSQeiTScmuyLU1Ho15epnkKi0fkc87c3mebyi8RW1wErxE9CUfjUZDHPN4Auwe3IAV4FMih5ZACSPGoDiWBX4yFAqHh1XKCteb1AqhUsAlcIRSjGoAtUxzNvzdrBACZHfI-EofYEvfUfRg-QGhEGAgRQ1Jw27mYm2SmuZc+ZmxfjthXiUOsAANOqvyvVtevvW9wTkaSLRvSL62ug1YAMz0JoTphFuBgemIuh7ihB7+qeJ4hsEl6RjSpoYKIC5bFBsAkh8MI-hRsG1kBjZgJKALAkqbGwAQI7Ft2-BtMWyqKlwCrwKgrG-LA0DApAbT5EgGBQpOxpKMR5oTJakFLoWPFYNWAAsTD4RSkS-suyAAeIjIGKZ0FgnBjDZppb7oFRWAAFq0VpsBFiWulggZgFCkcwFZG4+QAOR8NAcrwBJoDVDAvbIBoqCTICcXWgOngKvsnDtFgej4gAsjFxHrs6-j2E4GDBLIODJYCTSgCVbywGVNmUR+9wNaA3TNaVMYYLWlY+aW3X5PFtgtW1g3De+pKfuNgLcFNA0YnNLkfjCS0sKtrXtRRrlufcQA)

### 💻 Code

```ts
// Some helpers
type DropFirst<T extends any[]> = T extends [infer _, ...infer T] ? T : never
type Last<T extends any[]> = T extends [...infer _, infer L] ? L : never

// Some setup
type TestFunc = (a: string, b: (x: number, y: string) => void) => void
type TestParams = Parameters<TestFunc>

// Works without a template.  Yields [z: Error, y: string]
type TestTypeWorks =
  [...[z: Error], ...DropFirst<Parameters<Last<TestParams>>>]

// Breaks when used with a template?  Yields [Error, string]
type TestTypeTpl<T extends [...any[], (...args: any) => any]> =
  [...[z: Error], ...DropFirst<Parameters<Last<T>>>]
type TestTypeFails = TestTypeTpl<TestParams>

// Individual pieces work fine
// Yields [y: string]
type TestTypeTpl2<T extends [...any[], (...args: any) => any]> =
  [...DropFirst<Parameters<Last<T>>>]
type TestTypeWorksX = TestTypeTpl2<TestParams>

// Yields [z: Error]
type TestTypeTpl3<_ extends [...any[], (...args: any) => any]> =
  [...[z: Error]]
type TestTypeWorksY = TestTypeTpl3<TestParams>

// But then they fail again when we put them together. Yields [Error, string]
type TestTypeFailsTpl4<T extends [...any[], (...args: any) => any]> =
  [...TestTypeTpl2<T>, ...TestTypeTpl3<T>]
type TestTypeWorksZ = TestTypeFailsTpl4<TestParams>

```

### 🙁 Actual behavior

The tuples lost their labels.

### 🙂 Expected behavior

The tuples should still have their labels.

**Comment by brett19** [comment]
This might be a dupe of #42106 or #39941.  There is an implementation of Append which definitely does work around this issue there though, but I can't seem to figure out what's special about my instance that makes the names fail to be propagated.

```
type Append<A, B> = A extends [...infer Params] ? [...Params, ...(B extends [...infer Params2] ? Params2 : [])] : never;
```

**Comment by RyanCavanaugh** [comment]
Preserving tuple labels through transforms is "best effort" not guaranteed to work in call cases. If it's an easy fix we'd take a PR but don't want to do anything perf-impacting for a nice-to-have.

**Comment by DanielRosenwasser** [comment]
This is now fixed in TypeScript 5.2, probably by #53356 (see [the blog post](https://devblogs.microsoft.com/typescript/announcing-typescript-5-2/#named-and-anonymous-tuple-elements)).
